### PR TITLE
Find boundary

### PR DESCRIPTION
**Pull request by Jianshu-Hu**
Major feature: create function to search the task space and get the costs 
This is just a

### CODE REVIEW

File: examples/goldilocks_models/find_boundary.cc [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+  }
+  cout<<"cost_threshold "<<cost_threshold<<endl;
+
+  // create components for each dimension used to decide the direction
+  // considering we can only visualize 2D landscape, we only search within
+  // 2D space and fix other dimension
+  if(robot_option==0){
+    search_setting.SetExtendComponents("stride length",5,{0,-1,0.5,0.5,1});
+    search_setting.SetExtendComponents("ground incline",5,{0,-1,0.5,0.5,1});
+    search_setting.SetExtendComponents("velocity",1,{0});
```
Comment (yminchen) [inline examples/goldilocks_models/find_boundary.cc:674]: It doesn't look like we need the setter function `SetExtendComponents`. We can create "elements_" and pass it into the constructor at https://github.com/DAIRLab/dairlib/pull/171/files#diff-25ea7bf0927e455b7265fbe2b0032473R620 ?

File: examples/goldilocks_models/find_boundary_utils/visualize_gait_MBP.cc [hunk 2]
```diff
@@ -0,0 +1,318 @@
+#include <gflags/gflags.h>
```
[the remark this answers is not in the document]
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary_utils/visualize_gait_MBP.cc:1]: I made some changes to this file, but I didn't remember now. One of the changes is changing the directory to read data.

[the remark this answers is not in the document]
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary_utils/visualize_gait_MBP.cc:1]: I think I can remove this and make some changes to the `find_models/visualize_gait_MBP.cc` to make it work with find_boundary.

File: examples/goldilocks_models/find_boundary.cc [hunk 3]
```diff
@@ -0,0 +1,772 @@
+//
+// Created by jianshu on 5/20/20.
+//
```
[the remark this answers is not in the document]
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary.cc:3]: Sure, It is automatically created by CLion while I create a new file.

**Comment by Jianshu-Hu** [comment]
Finished changing the code as requested. I also add the python files used for plotting. It is the original version without applying the feature we talked on Tuesday. I will add the method of plotting triangular region later.

File: examples/goldilocks_models/find_boundary_utils/compare_boundary.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
+import matplotlib.pyplot as plt
+import matplotlib
+import numpy as np
+from mpl_toolkits.mplot3d import Axes3D
+from scipy.linalg import solve
+import torch
+from torch import tensor
```
Comment (yminchen) [inline examples/goldilocks_models/find_boundary_utils/compare_boundary.py:7]: Shouldn't import the libraries that we are not using

File: examples/goldilocks_models/find_boundary_utils/plot_landscape.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+if robot_option == 1:
+    min_gi1 = -0.325
+    max_gi1 = 0.275
+    min_sl1 = 0.1425
+    max_sl1 = 0.4425
+
+    min_gi2 = -0.175
+    max_gi2 = 0.125
+    min_sl2 = 0.2175
+    max_sl2 = 0.3675
```
Comment (yminchen) [inline examples/goldilocks_models/find_boundary_utils/plot_landscape.py:36]: ditto

File: examples/goldilocks_models/find_boundary_utils/plot_landscape.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+# large range
+if plot_large_range == 1:
+    x_line = np.array([min_sl1, max_sl1, max_sl1, min_sl1, min_sl1])
+    y_line = np.array([min_gi1, min_gi1, max_gi1, max_gi1, min_gi1])
+    ax.plot(x_line, y_line, 'black', linewidth=3)
+# small range
+if plot_small_range == 1:
+    x_line2 = np.array([min_sl2, max_sl2, max_sl2, min_sl2, min_sl2])
+    y_line2 = np.array([min_gi2, min_gi2, max_gi2, max_gi2, min_gi2])
+    ax.plot(x_line2, y_line2, 'black', linewidth=3)
```
Comment (yminchen) [inline examples/goldilocks_models/find_boundary_utils/plot_landscape.py:139]: It looks like we are plotting two boxes here? Shouldn't we only plot one box since we are not comparing two cost landscape?

File: examples/goldilocks_models/find_boundary_utils/compare_boundary.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+robot_option = 1
+file_dir = '/Users/jason-hu/'
+
+# optimization setting
+if robot_option == 0:
+    min_gi1 = -0.325
+    max_gi1 = 0.275
+    min_sl1 = 0.0925
+    max_sl1 = 0.3925
```
[the remark this answers is not in the document]
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary_utils/compare_boundary.py:19]: In the` initial guess.cc`, the scale was decided by the max and min value of the task space. Here I didn't figure out a similar method to decide them instead of hard-coding.

File: examples/goldilocks_models/find_boundary_utils/compare_boundary.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    dir1 = file_dir+'dairlib_data/find_boundary/' + robot + 'robot_' + str(robot_option) + \
+           '_nominal/'
+    dir2 = file_dir+'dairlib_data/find_boundary/' + robot + 'robot_' + str(robot_option) + \
+           '_initial_model/'
+    dir3 = file_dir+'dairlib_data/find_boundary/' + robot + 'large_task_space/' + 'robot_' + str(robot_option) + \
+           '_large_task_space_iter1000/'
+    dir4 = file_dir+'dairlib_data/find_boundary/' + robot + 'large_task_space/' + 'robot_' + str(robot_option) + \
+           '_large_task_space_iter2000/'
+    dir5 = file_dir+'dairlib_data/find_boundary/' + robot + 'large_task_space/' + 'robot_' + str(robot_option) + \
+           '_large_task_space_iter3000_with_scaling/'
```
[the remark this answers is not in the document]
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary_utils/compare_boundary.py:50]: This is just what I use to name different data set. This data set was generated using snopt scaling so I named it "with scaling".

File: examples/goldilocks_models/find_boundary_utils/plot_landscape.py (hunk 6 above)
Comment (Jianshu-Hu) [reply examples/goldilocks_models/find_boundary_utils/plot_landscape.py:139]: We can choose which box to plot by setting the falg `plot_xxx_range`. This function is also used for comparing the results of optimizing over different task space (Eg. normalizing the cost from the large one with the cost from the small one).

**Comment by Jianshu-Hu** [comment]
1. Generalize the search function by using a recursion.
2. Make changes to codes following the comments
